### PR TITLE
Add bounds to range and autorange of cartesian, gl3d and radial axes

### LINKED ISSUE #400: Setting only min or max values for axis and not a range.
Dear all

Many thanks for all your wonderful efforts with Plotly

I just wanted to ask if it might be possible to tweak the range function a little

currently an axis range can be set in the layout like this ...
```
layout = go.Layout(
xaxis=dict(
range=[2, 5]
),
```
in some cases it would be desirable to only specify a min value.
Plotly can then work out the max value automatically.

Or perhaps a user would want to set a max value and have a plotly work out the min value

Either-way it would be nice (for me at least) to only specify one of the range limits and not both. However I can see why this might not be a highly required function and I'm sure there are lots of other things to do first

matplotlib offers this function ...
```
plt.xlim(2,5) #sets a range
plt.gca().set_xlim(xmin=2) # sets only the lower x value and allows computer to work out upper value
plt.gca().set_ylim(ymax=50) # sets only the lower y value and allows computer to work out upper value
```

### PR DESCRIPTION
**Pull request by archmoj**
Resolves #400.
@plotly/plotly_js

### CODE REVIEW

**Comment by alexcjohnson** [comment]
After chatting with @archmoj last week and playing with this, I think I understand the attributes implemented here:
- `autorange(min|max)` - the smallest and largest values that are allowed to be in the results of autoranging this axis. So if these are outside the autorange we calculate they have no effect
- `range(min|max)` - the smallest and largest values that are allowed to be in the range EVER. If you try to drag beyond this, or `relayout` beyond this, we replace either bound that goes outside this range.

Both of these are useful, but I think this is missing an important one, perhaps the first one people will think of in this vein: a specific value to use for one or the other end of autorange, for example "no matter the data, I want the bottom of the range to be 50, but I want Plotly to choose the top of the range." This would almost be a generalization of `rangemode: 'tozero'`, except that even if the data go past zero, the autorange would end at exactly zero.

So I think we need three pairs of attributes, not just two. What if we put them all in a new container:
```
rangebounds: {
    automin / automax: use these values exactly for autorange
    softmin / softmax: do not let autorange go beyond these values. Ignore if corresponding auto is provided
    hardmin / hardmax: do not let range (auto, explicit, or dynamic) go beyond these values
}
```
`rangebounds.softmin` is what you currently have as `autorangemin`, and `rangebounds.hardmin` is what you have as `rangemin`. `rangebounds.automin` is new behavior.

And two things I notice playing with the PR as it stands today:
- If one end of autorange is restricted, currently this has no impact on the other end, but it should. For example if there are scatter markers we pad by the marker size plus 5%, but this padding is a smaller data value if the other end of the range is clipped, and bigger if the other end is extended.
- Make sure the dynamic restrictions work right when you drag. Currently when you drag the trace keeps moving, but the axis stays put. Furthermore, if you drag both ends of the axis (either dragging the center of the axis or panning the whole subplot) and one gets clipped, I think the other end should stop too, rather than starting to zoom at this point.

**Comment by nicolaskruchten** [comment]
I don't understand "use these values exactly for autorange" ... ? If we say "use these" then what is "auto"?

**Comment by nicolaskruchten** [comment]
I think a little table like this would help a little:

| range min/max (in) | rangemode | real data min/max | auto min/max | soft min/max | hard min/max | range min/max (out) |
| -| -| -| -| -| -| -|

I assume the idea is that one could set the "in" here by e.g. panning or providing `range: [min, max]` and then the library would chunk through some calcs that use all of the above as inputs and overwrite `range` with "out"?

**Comment by nicolaskruchten** [comment]
I'll have some longer thoughts shortly but here are some assorted intuitions:
- "auto" to me suggests the output of some automated procedure, which is correct for Plotly.js but I think that it's not obvious to most users when "auto" kicks in. It's basically "whenever `range` is unset" and "sometimes when double-clicking" right? I also think that most people will not intuit that pan/zoom overwrites `range`, such that double-clicking would re-invoke auto-range.
- "min"/"max" have connotations of "minimum actual" i.e. "bottom" or "lower" but also "minimum possible".
- Taken together, "automin" to me suggests "the smallest possible value that an automated procedure would output" not "the fixed bottom value", that is to say what you call "softmin".
- I'd also add that as a user, I'd expect that since setting `range` to `null` (and `[null, null]` also?) causes auto-ranging to happen on both values, therefore setting `range` to `[3, null]` should cause auto-ranging on the top value by default so I'd recommend having this `[3, null]` case automatically set the appropriate values of the new attributes if possible!

**Comment by nicolaskruchten** [comment]
In reaction to your new API proposal, I think that `min`/`clipmin` speaks to this ambiguity around "bottom" vs "minimum output of procedure" :)

**Comment by nicolaskruchten** [comment]
I love the `include` option!

**Comment by nicolaskruchten** [comment]
Here's the spreadsheet I'm imagining as the test grid for this feature btw: https://docs.google.com/spreadsheets/d/1DMyGYDSiNoemS78zKjdf_MKKGfSxm3iLQH4r64q-BNg/edit#gid=0

I have some questions around how these interact, and what happens if a min is greater than a max...

**Comment by chriddyp** [comment]
With respect the original issue, could a simpler API just be `xaxis.range = [5, None]` and `xaxis.range = [None, 10]`? If specified `None`, then we base it off of the data. That way users don't have to learn a new attribute and it matches the current shape of things.

**Comment by alexcjohnson** [comment]
> With respect the original issue, could a simpler API just be `xaxis.range = [5, null]` and `xaxis.range = [null, 10]`? If specified `null`, then we base it off of the data. That way users don't have to learn a new attribute and it matches the current shape of things. (AJ edited `None` -> `null`, we're in JS here 😉)

I do love the simplicity of this API, and it would let us get rid of `autorangeoptions.min/max` but we'd keep `autorangeoptions.clipmin/max` and `rangemin/max`. Or maybe we call both of these `min/maxallowed`? In many ways I like that more: consistent names, makes it clear these are limits (`rangemin` could be taken to mean "the min value is only allowed to be that" rather than "the min is not allowed to go lower than that"). The only thing I don't like is that `xaxis.minallowed` doesn't explicitly say `range` so it could be taken to mean "no data to be plotted here can be less than that" - but I think it's better to keep these short and consistent with the names inside `autorangeoptions`.

What happens though if we later change the data? We write the computed range back to `layout.xaxis.range` so we've lost the information that only one end should autorange. There's also a question of how to handle reversed axes.

So what if we create 4 new `axis.autorange` values to go along with these situations:
- `xaxis.range = [5, null]` would cause a default `xaxis.autorange = 'max'` ie "only autorange the max end of the range" but if you want 5 to be the max of a reversed range you can also set `xaxis.autorange = 'min reversed'`
- `xaxis.range = [null, 10]` would cause a default `xaxis.autorange = 'min'` ie "only autorange the min end of the range" but if you want 10 to be the min of a reversed range you can also set `xaxis.autorange = 'max reversed'`

Next question is what happens if the viewer doubleclicks, or zooms/pans and then doubleclicks - seems like mostly this should do a full autorange, ie it sets `autorange=true` (and then by the current default behavior the second doubleclick takes you back to the initial range), but if you wanted autorange still limited or extended on either end you could set `autorangeoptions` to achieve this. Or if you always want doubleclick to give the initial range you can set `config.doubleClick='reset'`

So the new API would be:
```
minallowed: min value allowed in range. Even explicitly setting range:[a,b] can't go past this.
maxallowed: max value allowed in range. Even explicitly setting range:[a,b] can't go past this.
    must be more than minallowed
range: allowed to leave either or both elements `null`, this impacts the default `autorange`
autorange: 'min', 'max', 'min reversed', 'max reversed' as well as existing true, false, 'reversed'
    if 'min' or 'max' is in the value, only autorange that end, use existing `range` entry for the other end
    if range[0] is missing, 'min' is the default and false, 'max', and 'min reversed' are prohibited
    if range[1] is missing, 'max' is the default and false, 'min', and 'max reversed' are prohibited
autorangeoptions: {
    minallowed: do not let min autorange go below this. Ignore if autorange = false, 'max', 'max reversed',
        or if it's <axis.minallowed or >axis.maxallowed
    maxallowed: do not let max autorange go above this. Ignore if autorange = false, 'min', 'min reversed',
        or if it's <axis.minallowed or >axis.maxallowed
    include: number or [num1, num2], ensure the autorange min <= and max >= these values,
        ignore any num outside either set of min/maxallowed constraints
}
```

**Comment by LiamConnors** [comment]
@alexcjohnson @archmoj I'm not sure I understand the distinction between `maxallowed` and `clipmax` and (`minallowed` and `clipmin`) here. In what situation would you use one over the other?

Trying these examples in Plotly.py, I see the same behaviour, including when panning.

```
import plotly.express as px
df = px.data.tips()
fig = px.scatter(df, x="total_bill", y="tip", color="sex")

fig.update_xaxes(
autorangeoptions=dict(
    clipmax=40
)
)

fig.show()
```

```
import plotly.express as px
df = px.data.tips()
fig = px.scatter(df, x="total_bill", y="tip", color="sex")

fig.update_xaxes(
autorangeoptions=dict(
    maxallowed=40
)
)

fig.show()
```

**Comment by archmoj** [comment]
> 

@LiamConnors what is the maximum value within your dataset? 
If that's less than 40, you can get similar results. To see the difference you may use a value less than the maximum of your dataset.

**Comment by LiamConnors** [comment]
> > 
> 
> @LiamConnors what is the maximum value within your dataset? If that's less than 40, you can get similar results. To see the difference you may use a value less than the maximum of your dataset.

In this case, the maximum on the x axis is 50.81. 
It would look like this without anything set

image

**Comment by archmoj** [comment]
> > > 
> > 
> > 
> > @LiamConnors what is the maximum value within your dataset? If that's less than 40, you can get similar results. To see the difference you may use a value less than the maximum of your dataset.
> 
> In this case, the maximum on the x axis is 50.81. It would look like this without anything set
> 
> image

It looks like those attributes are not set on the axis. 
@LiamConnors Could you possibly make a codepen from the figure?

**Comment by LiamConnors** [comment]
> It looks like those attributes are not set on the axis. @LiamConnors Could you possibly make a codepen from the figure?

Thanks @archmoj. I made two codepens for the two examples:

https://codepen.io/Liam-Connors/pen/QWJVYyZ?editors=1010

https://codepen.io/Liam-Connors/pen/mdQGvNG?editors=1010

**Comment by archmoj** [comment]
> 

Thanks @LiamConnors for the codepens. 
Both work as expected on the `plotly.js` side. 
So the issue should be on the `plotly.py` side. Right?

**Comment by LiamConnors** [comment]
> > 
> 
> Thanks @LiamConnors for the codepens. Both work as expected on the `plotly.js` side. So the issue should be on the `plotly.py` side. Right?

I am more wondering what the two examples are doing differently and why you would set `maxallowed=40` instead of `clipmax=40`. I don't know if the example I've given has the data to demonstrate the difference, but in the codepens they don't seem to be doing anything differently.

**Comment by archmoj** [comment]
> > > 
> > 
> > 
> > Thanks @LiamConnors for the codepens. Both work as expected on the `plotly.js` side. So the issue should be on the `plotly.py` side. Right?
> 
> I am more wondering what the two examples are doing differently and why you would set `maxallowed=40` instead of `clipmax=40`. I don't know if the example I've given has the data to demonstrate the difference, but in the codepens they don't seem to be doing anything differently.

Using 40 with this dataset we get similar results which is correct. 
But with 60 you get different results which illustrates the difference between the two attributes.

**Comment by LiamConnors** [comment]
> > > > 
> > > 
> > > 
> > > Thanks @LiamConnors for the codepens. Both work as expected on the `plotly.js` side. So the issue should be on the `plotly.py` side. Right?
> > 
> > 
> > I am more wondering what the two examples are doing differently and why you would set `maxallowed=40` instead of `clipmax=40`. I don't know if the example I've given has the data to demonstrate the difference, but in the codepens they don't seem to be doing anything differently.
> 
> Using 40 with this dataset we get similar results which is correct. But with 60 you get different results which illustrates the difference between the two attributes.

Ah okay. That makes sense to me now. Thanks!

**Comment by archmoj** [comment]
@LiamConnors Also see Alex's `new API suggestion` here: https://github.com/plotly/plotly.js/pull/6547#issuecomment-1513302220

File: src/plots/cartesian/layout_attributes.js [hunk 1]
```diff
@@ -272,17 +272,68 @@ module.exports = {
...
             'Determines whether or not the range of this axis is',
             'computed in relation to the input data.',
             'See `rangemode` for more info.',
-            'If `range` is provided, then `autorange` is set to *false*.'
+            'If `range` is provided, then `autorange` is set to *false*.',
+            'Using *min* applys autorange only to set the minimum.',
+            'Using *max* applys autorange only to set the maximum.',
+            'Using *min reversed* applys autorange only to set the minimum on a reversed axis.',
+            'Using *max reversed* applys autorange only to set the maximum on a reversed axis.',
+            'Using *reversed* applys autorange on both ends and reverse the axis direction.',
```
Comment (LiamConnors) [inline src/plots/cartesian/layout_attributes.js:288]:
```suggestion
            'If `range` is provided, then `autorange` is set to *false*.',
            'Using *min* applies autorange only to set the minimum.',
            'Using *max* applies autorange only to set the maximum.',
            'Using *min reversed* applies autorange only to set the minimum on a reversed axis.',
            'Using *max reversed* applies autorange only to set the maximum on a reversed axis.',
            'Using *reversed* applies autorange on both ends and reverses the axis direction.',
```

**Comment by archmoj** [comment]
> 

@LiamConnors good catch it looks like `autorangeoptions` is not used when range is set partially.
I also simplified the codepen [here](https://codepen.io/MojtabaSamimi/pen/qBQMwWL?editors=0010) so that it could be easier to play with it.

**Comment by LiamConnors** [comment]
Thanks @archmoj. I retested the previous issue I noticed and it's now working as expected.

One last issue I note is when using this plotly.express example:

```
import plotly.express as px
df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length", facet_col="species")
fig.update_xaxes(range=[1.5, None])
fig.update_yaxes(range=[3, None])

fig.show()
```

It doesn't calculate the autorange as I would expect:

image

Here's the plotly.js codepen:

https://codepen.io/Liam-Connors/pen/LYXXWJo?editors=1010

**Comment by archmoj** [comment]
> Thanks @archmoj. I retested the previous issue I noticed and it's now working as expected.
> 
> One last issue I note is when using this plotly.express example:
> 
> ```
> import plotly.express as px
> df = px.data.iris()
> 
> fig = px.scatter(df, x="sepal_width", y="sepal_length", facet_col="species")
> fig.update_xaxes(range=[1.5, None])
> fig.update_yaxes(range=[3, None])
> 
> fig.show()
> ```
> 
> It doesn't calculate the autorange as I would expect:
> 
> image
> Here's the plotly.js codepen:
> 
> https://codepen.io/Liam-Connors/pen/LYXXWJo?editors=1010

Good catch @LiamConnors  :pray:
Now fixed in the PR.

**Comment by LiamConnors** [comment]
Thanks @archmoj. Looks good to me know. And I don't see any further issues.

File: src/plots/cartesian/layout_attributes.js [hunk 2]
```diff
@@ -272,17 +272,68 @@ module.exports = {
...
+        values: [true, false, 'reversed', 'min reversed', 'max reversed', 'min', 'max'],
         dflt: true,
         editType: 'axrange',
         impliedEdits: {'range[0]': undefined, 'range[1]': undefined},
         description: [
             'Determines whether or not the range of this axis is',
             'computed in relation to the input data.',
             'See `rangemode` for more info.',
-            'If `range` is provided, then `autorange` is set to *false*.'
+            'If `range` is provided, then `autorange` is set to *false*.',
```
Comment (alexcjohnson) [inline src/plots/cartesian/layout_attributes.js:283]: this isn't necessarily true anymore 😎

File: src/plots/cartesian/layout_attributes.js [hunk 3]
```diff
@@ -272,17 +272,68 @@ module.exports = {
...
+            'Using *min* applies autorange only to set the minimum.',
```
Comment (LiamConnors) [inline src/plots/cartesian/layout_attributes.js:284]:
```suggestion
            'If `range` is provided and it has a value for both the',
            ' lower and upper bound, `autorange` is set to *false*.',
            'Using *min* applies autorange only to set the minimum.',
```

File: src/plots/cartesian/layout_attributes.js [hunk 4]
```diff
@@ -272,17 +272,69 @@ module.exports = {
...
         dflt: true,
         editType: 'axrange',
         impliedEdits: {'range[0]': undefined, 'range[1]': undefined},
         description: [
             'Determines whether or not the range of this axis is',
             'computed in relation to the input data.',
             'See `rangemode` for more info.',
-            'If `range` is provided, then `autorange` is set to *false*.'
+            'If `range` is provided and it has a value for both the',
+            ' lower and upper bound, `autorange` is set to *false*.',
```
Comment (LiamConnors) [inline src/plots/cartesian/layout_attributes.js:284]:
```suggestion
            'lower and upper bound, `autorange` is set to *false*.',
```
Sorry @archmoj, I guess this space should go for the scheme to generate with only one.

**Comment by alexcjohnson** [comment]
@archmoj if I set one end of the range and ask for normal order, the behavior now is perfect AFAICT, ie either of these a doubleclick alternates between the original range and a full autorange, and if I first pan/zoom and then doubleclick, it goes to the original range first, then to a full autorange.:
```js
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[5.5,null]}})
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[null, 5.5]}})
```
And if I get the reversed setting precisely correct, the behavior is also correct, with axis reversed and doubleclick toggling between initial state and full autorange:
```js
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[null, 5.5], autorange:'max reversed'}})
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[5.5, null], autorange:'min reversed'}})
```
However, if I instead add `autorange: 'reversed'` (which I would have expected would be interpreted the same as the two cases above), then a doubleclick does nothing (until I pan/zoom, then doubleclick takes me back to the original as desired, but a second doubleclick should then go to full autorange).
```js
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[null, 5.5], autorange:'reversed'}})
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[5.5, null], autorange:'reversed'}})
```

Also if I pick the wrong reversed option:
```js
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[null, 5.5], autorange:'max'}})
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[5.5, null], autorange:'min'}})
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[null, 5.5], autorange:'min reversed'}})
Plotly.newPlot(gd,[{y:[4,5,6,7]}],{yaxis:{range:[5.5, null], autorange:'max reversed'}})
```
I think would expect the `autorange` setting to be ignored, except maybe for whether it's reversed or not. And that's how it looks initially (the provided end of the range is respected and the other end is autoranged, even though the setting asked for the opposite), but after that a doubleclick takes me to full autorange and then with further doubleclicks we stay there, rather than going back to the initial range.

**Comment by archmoj** [comment]
Thanks for the review @alexcjohnson.
For those invalid `range` and `autorange` combinations now I added validation and tests so that we don't accept `range` and set `autorange` to *true*.

**Comment by archmoj** [comment]
> Getting much closer! There's still some weird behavior if you set a partial range and `autorange: true` (it keeps the specified end, but on double-click does not fully autorange)

Good catch. Addressed in [30a692e](https://github.com/plotly/plotly.js/pull/6547/commits/30a692e1d1bf3eb3a6a8266011b11b570606e43a).

**Comment by archmoj** [comment]
> There's still some weird behavior ... `autorange: 'reversed'` (it drops BOTH the reversing and the partial range and gives you full non-reversed autorange).

I think this is a desired behavior to ignore ambiguous `range` and `autorange` combinations and set `autorange` to *true*.

**Pull request merged**